### PR TITLE
Add full support for contrasts to Formulaic

### PR DESCRIPTION
**Pull request by matthewwardrop**
This patch set is largely complete, but lacks documentation and a few more unit tests for various edge cases. Nevertheless, everything should work pretty robustly as is.

As of this PR, you can use arbitrary contrasts in a formula, e.g.: `y ~ C(A, contr.helmert)`, or `y ~ C(A, contr.treatment("base"))`, or `y ~ C(A, {"coding": [...], ...})`, etc.

- [ ] Add documentation, type annotations, etc.

### CODE REVIEW

**Comment by matthewwardrop** [comment]
@bashtage fyi. Let me know if this support for contrast matrices is insufficient for `statsmodels` use-cases.

**Comment by bashtage** [comment]
I'll take a look, thanks.

File: formulaic/utils/sparse.py [hunk 1]
```diff
@@ -1,37 +1,43 @@
...
+    series: Iterable, levels: Optional[Iterable[str]] = None, drop_first: bool = False
 ) -> spsparse.csc_matrix:
     """
     Categorically encode (via dummy encoding) a `series` as a sparse matrix.
 
     Args:
         series: The iterable which should be sparse encoded.
+        levels: The levels for which to generate dummies (if not specified, a
+            dummy variable is generated for every level in `series`).
         reduced_rank: Whether to omit the first column in order to avoid
```
Comment (bashtage) [inline formulaic/utils/sparse.py:18]: You removed reduced_rank and now have drop_first.

**Comment by bashtage** [comment]
Ok, so I'll jut go with a noob question. Suppose my model is `y ~ 1 + x1 + x2 + x3`.  How do I get the linear restriction matrix for the contrast `x2 + x3 = 1`?  What about `x1=0; x2 + x3 = 1`?

**Comment by matthewwardrop** [comment]
Hi @bashtage ! Apologies, this PR doesn't add support for linear constraints; that work is separate. I suspect your message in the other issue thread was a typo, then? You wrote "contrasts" bit perhaps meant constraints?

I'll bump the priority of that work too.

**Comment by bashtage** [comment]
I meant in this way: https://en.wikipedia.org/wiki/Contrast_(statistics) . It is what statsmodels calls these things (I don't like the name, but...).  You are right that I'm mostly looking for https://patsy.readthedocs.io/en/latest/API-reference.html#linear-constraints 

Sorry for the confusion.

**Comment by matthewwardrop** [comment]
Huh... interesting. So actually, that's canonically the sense in which I'm referring to "contrasts" as well, but here in the sense of using them to encode a categorical variable into a full rank matrix. It didn't occur to me to think of them as the same thing, since I thought that the constraints (being tracked in #38) were acting on columns of the model matrix (rather than levels of a category), and it's not clear to me how they sum to zero. I thought linear constraints would be anything of form `Ax = b` with `A` a matrix of coefficients for combinations of the features of a model matrix `x`, and `b` a vector of constants. Am I missing something?

**Comment by bashtage** [comment]
Both features are very useful - the dummy coding for full compat with patsy, and constraints for hypothesis testing.  Thanks.
